blog/_posts/2025-04-04-Traffic-Routing-Demystified-Best-practices-for-AKS-Load-Balancers.md

```diff
@@ -0,0 +1,141 @@
+
+- **For HTTP/1.1 Connections**:
+    The server should include a `Connection: close` header in its response for all active and new incoming requests. This informs clients not to reuse the connection and allows idle connections to be closed gracefully.
+    **Use Case**: Applications serving REST APIs or web traffic where clients rely on persistent connections for performance optimization.
+
+> **Note**: In HTTP/1.1, there is a potential race condition where the server might close an idle connection at the same time the client sends a new request. In such cases, the client must handle this scenario by retrying the request on a new connection.
+
+- **For HTTP/2 Connections**:
+    The server should send a `GOAWAY` frame to notify clients that the connection is being closed. This allows clients to gracefully terminate the connection and retry requests on a new connection if necessary.
+    **Use Case**: Applications using gRPC or HTTP/2 for high-performance communication between services or with external clients.
+
+### Preventing New Requests to an Unhealthy Pod
+To avoid routing new requests to a pod that is in the process of shutting down, it is important to manage its health status effectively. The below image shows the timeline for a pod receiving a `TERM` signal and gracefully shutting down without impact on external traffic:
+
+![Preventing New Requests to Unhealthy Pods](./AKS/blog/assets/images/preventingnewrequeststoanunhealthypod.png)
+
+- **Immediate Health Check Response**:
+    As soon as a pod is marked for deletion, its `healthCheckNodePort` should start returning a 500 error. This signals to external load balancers that the pod is no longer healthy and should not receive new traffic.
+
+- **Load Balancer Probe Delay**:
+    External load balancers may take a few seconds to detect the unhealthy status of a pod. During this time, the pod might still receive traffic.
+
+To ensure your pods follow a similar timeline to gracefully shutdown, make sure to assess the following:
+
+1. **Delay Termination**:
+     After receiving the `TERM` signal, the application should wait for at least 10 seconds before proceeding with shutdown tasks (helps address the Load balancer probe delay). This can be achieved using a `preStop` hook in Kubernetes.
+
+    > **Note**: When using annotations `service.beta.kubernetes.io/azure-load-balancer-health-probe-interval` and/or `service.beta.kubernetes.io/azure-load-balancer-health-probe-num-of-probe`, consider changing the wait time to cover the annotation's needs (see [documentation](https://cloud-provider-azure.sigs.k8s.io/topics/loadbalancer/))
+
+2. **Announce Readiness as `false`**:
+     The application should update its readiness probe to indicate it is no longer ready to serve traffic. This allows Kubernetes to stop routing new requests to the pod.
```

Comment:
In the ExternalTrafficPolicy=Local case, the application must announce readiness false **after** the previous delay.  The purpose of this is to ensure that if there are more than one application pods on the same node, that the application doesn't keep on receiving connection requests at this point.